parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/ColumnEncryptor.java

```diff
@@ -0,0 +1,317 @@
+      }
+
+      writer.endBlock();
+      store = reader.readNextRowGroup();
+      blockId++;
+    }
+  }
+
+  private void processChunk(ColumnDescriptor descriptor, ColumnChunkMetaData chunk, TransParquetFileReader reader, ParquetFileWriter writer,
+                            Set<ColumnPath> encryptPaths, int blockId, int columnId, String createdBy) throws IOException {
+    reader.setStreamPosition(chunk.getStartingPos());
+    writer.startColumn(descriptor, chunk.getValueCount(), chunk.getCodec());
+    processPages(reader, chunk, writer, createdBy, blockId, columnId, encryptPaths.contains(chunk.getPath()));
+    writer.endColumn();
+  }
+
+  private void processPages(TransParquetFileReader reader, ColumnChunkMetaData chunk, ParquetFileWriter writer,
+                            String createdBy, int blockId, int columnId, boolean encrypt) throws IOException {
+    int pageOrdinal = 0;
+    EncryptorRunTime encryptorRunTime = new EncryptorRunTime(writer.getEncryptor(), chunk, blockId, columnId);
+    DictionaryPage dictionaryPage = null;
+    long readValues = 0;
+    ParquetMetadataConverter converter = new ParquetMetadataConverter();
+    OffsetIndex offsetIndex = reader.readOffsetIndex(chunk);
+    reader.setStreamPosition(chunk.getStartingPos());
+    long totalChunkValues = chunk.getValueCount();
+    while (readValues < totalChunkValues) {
+      PageHeader pageHeader = reader.readPageHeader();
+      byte[] pageLoad;
+      switch (pageHeader.type) {
+        case DICTIONARY_PAGE:
+          if (dictionaryPage != null) {
+            throw new IOException("has more than one dictionary page in column chunk");
+          }
+          //No quickUpdatePageAAD needed for dictionary page
+          DictionaryPageHeader dictPageHeader = pageHeader.dictionary_page_header;
+          pageLoad = processPayload(reader, pageHeader.getCompressed_page_size(), encryptorRunTime.getDataEncryptor(), encryptorRunTime.getDictPageAAD(), encrypt);
+          writer.writeDictionaryPage(new DictionaryPage(BytesInput.from(pageLoad),
+                                        pageHeader.getUncompressed_page_size(),
+                                        dictPageHeader.getNum_values(),
+                                        converter.getEncoding(dictPageHeader.getEncoding())),
+            encryptorRunTime.getMetaDataEncryptor(), encryptorRunTime.getDictPageHeaderAAD());
+          break;
+        case DATA_PAGE:
+          if (encrypt) {
+            AesCipher.quickUpdatePageAAD(encryptorRunTime.getDataPageHeaderAAD(), pageOrdinal);
+            AesCipher.quickUpdatePageAAD(encryptorRunTime.getDataPageAAD(), pageOrdinal);
+          }
+          DataPageHeader headerV1 = pageHeader.data_page_header;
+          pageLoad = processPayload(reader, pageHeader.getCompressed_page_size(), encryptorRunTime.getDataEncryptor(), encryptorRunTime.getDataPageAAD(), encrypt);
```

Comment:
should be handled differently if `encrypt` is false

Reply:
Same as above.